source/Halibut/Transport/ConnectionManager.cs

```diff
@@ -14,21 +14,36 @@ public class ConnectionManager : IDisposable
         public bool IsDisposed { get; private set; }
 
         public IConnection AcquireConnection(IConnectionFactory connectionFactory, ServiceEndPoint serviceEndpoint, ILog log)
+        {
+            var openableConnection = GetConnection(connectionFactory, serviceEndpoint, log);
+            openableConnection.Item2(); // Since this involves IO, this should never be done inside a lock
+            return openableConnection.Item1;
+        }
+        
+        // Connection is Lazy instantiated, so it is safe to use. If you need to wait for it to open (eg for error handling, an openConnection method is provided)
+        // For existing open connections, the openConnection method does nothing
+        Tuple<IConnection, Action> GetConnection(IConnectionFactory connectionFactory, ServiceEndPoint serviceEndpoint, ILog log)
         {
             lock (activeConnections)
             {
-                var connection = pool.Take(serviceEndpoint) ?? CreateConnection(connectionFactory, serviceEndpoint, log);
-                AddConnectionToActiveConnections(serviceEndpoint, connection);
-
-                return connection;
+                var existingConnectionFromPool = pool.Take(serviceEndpoint);
+                var openableConnection = existingConnectionFromPool != null 
+                    ? Tuple.Create<IConnection, Action>(existingConnectionFromPool, () => { }) // existing connections from the pool are already open
+                    : CreateNewConnection(connectionFactory, serviceEndpoint, log);
+                AddConnectionToActiveConnections(serviceEndpoint, openableConnection.Item1);
+                return openableConnection;
             }
         }
 
-        IConnection CreateConnection(IConnectionFactory connectionFactory, ServiceEndPoint serviceEndpoint, ILog log)
+        Tuple<IConnection, Action> CreateNewConnection(IConnectionFactory connectionFactory, ServiceEndPoint serviceEndpoint, ILog log)
```

Comment:
Could you return `(IConnection connection, Action create)` instead?

Reply:
I opted not to do that right now because this requires a new dependency on `System.ValueTuple`. I'll ship it as is and come back to this

Reply:
Ah, no worries, I thought it already had that, but that must be Nevermore